server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/DeleteAction.java

```diff
@@ -0,0 +1,109 @@
+import org.sonar.db.component.BranchKeyType;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.db.permission.OrganizationPermission;
+import org.sonar.server.component.ComponentCleanerService;
+import org.sonar.server.user.UserSession;
+
+import static org.sonar.server.ws.KeyExamples.KEY_PROJECT_EXAMPLE_001;
+import static org.sonar.server.ws.WsUtils.checkFoundWithOptional;
+import static org.sonarqube.ws.client.projectbranches.ProjectBranchesParameters.ACTION_DELETE;
+import static org.sonarqube.ws.client.projectbranches.ProjectBranchesParameters.PARAM_PROJECT;
+import static org.sonarqube.ws.client.projectbranches.ProjectBranchesParameters.PARAM_BRANCH;;
+
+public class DeleteAction implements BranchWsAction {
+  private final DbClient dbClient;
+  private final UserSession userSession;
+  private final ComponentCleanerService componentCleanerService;
+
+  public DeleteAction(DbClient dbClient, UserSession userSession, ComponentCleanerService componentCleanerService) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+    this.componentCleanerService = componentCleanerService;
+  }
+
+  @Override
+  public void define(NewController context) {
+    WebService.NewAction action = context.createAction(ACTION_DELETE)
+      .setSince("6.6")
+      .setDescription("Delete a non-main branch of a project")
+      .setResponseExample(Resources.getResource(getClass(), "list-example.json"))
+      .setInternal(true)
+      .setPost(true)
+      .setHandler(this);
+
+    action
+      .createParam(PARAM_PROJECT)
+      .setDescription("Project key")
+      .setExampleValue(KEY_PROJECT_EXAMPLE_001)
+      .setRequired(true);
+    action
+      .createParam(PARAM_BRANCH)
+      .setDescription("Name of the branch to delete. Can't be the main branch of the project.")
+      .setExampleValue("branch1")
+      .setRequired(true);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn();
+    String projectKey = request.mandatoryParam(PARAM_PROJECT);
+    String branchKey = request.mandatoryParam(PARAM_BRANCH);
+
+    if (branchKey.isEmpty()) {
+      branchKey = null;
+    }
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      ComponentDto project = checkFoundWithOptional(dbClient.componentDao().selectByKey(dbSession, projectKey), "Project key '%s' not found", projectKey);
+
+      checkPermission(project);
+
+      BranchDto branch = checkFoundWithOptional(
+        dbClient.branchDao().selectByKey(dbSession, project.uuid(), BranchKeyType.BRANCH, branchKey),
+        "Branch '%s' not found for project '%s'", branchKey, projectKey);
+
+      if (branch.isMain()) {
+        throw new IllegalArgumentException("Only non-main branches can be deleted");
+      }
+      componentCleanerService.delete(dbSession, project);
+      response.noContent();
+    }
+  }
+
+  private void checkPermission(ComponentDto project) {
```

Comment:
Only admin of the project should be able to do this, so Please replace the content of this method by : 
userSession.checkComponentPermission(UserRole.ADMIN, project);